✨ [source-hubspot] Implement RFR for contacts_form_submissions, contact_merged_audit

brianjlai opened the pull request:
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7047

## What

There are 5 Hubspot streams that only support full refresh. I've implemented 3 of them:

- contacts_form_submissions
- contact_list_memberships
- contact_merged_audit

I did not implement the others because:
- `email_subscriptions`: This does not support any form of pagination and is therefore not RFR compatible
- `marketing_emails`: As per the [docs](https://legacydocs.hubspot.com/docs/methods/cms_email/get-all-marketing-emails), this supports filtering on `updated` or `created` and should just be converted to incremental which is better.

## How

Bumped the version to `0.87.0` which is the first version to have the RFR Python interfaces

To support RFR as per the design, the streams in question need to implement a version of `read_records()` which reads the prior page from the `stream_slice`, reads one page at a time, and updates state w/ the `next_page_token`.

The Hubspot streams being implemented use a form of cursor pagination where the response contains a pointer to the next contact (specified by the `vid`). And that `vid` is passed to the next request using the `vidOffset` parameter.

All of the full refresh streams derive from the same concept of fetching Contacts records and extracting a specific attribute list and returning them as records so I consolidated the implementation in the `ContactsAllBase` class.

I also added some mock server tests for each.

## Review guide

1. `source_hubspot/streams.py`
2. honestly the rest is just tests


## User Impact

Theoretically none, I did refactor `contact_merge_audit` to use the same flow as the others which was just cleaner, but now we should start seeing checkpoint messages emitted per-page during syncs

## Can this PR be safely reverted and rolled back?

No db or config migrations

- [x] YES 💚
- [ ] NO ❌

brianjlai commented:
Live tests ✅  against a connection: https://github.com/airbytehq/airbyte/actions/runs/9024913362/job/24799667426

maxi297 commented:
I'll check the PR tomorrow but note that `contact_list_memberships` will be move to semi-incremental next week

bazarnov commented on airbyte-integrations/connectors/source-hubspot/unit_tests/integrations/test_contacts_form_submissions.py:
```diff
@@ -0,0 +1,60 @@
+from .request_builders.streams import ContactsStreamRequestBuilder
+
+
+@freezegun.freeze_time("2024-05-04T00:00:00Z")
```
It's better to use the dynamic `end_date` freeze, like `pendulum.parse(<config.start_date>).substract(days=<x_days_to_substract>)`, having the 1 `start_date` anchor, instead of hardcoding the value.

bazarnov replied:
Also, these tests are similar except of the `self.mock_response` and `assert output.state_messages`, etc assertions, which make me think we can simplify the logic of these tests having 1 executable test_method, and bunch of inputs / outputs (parametrized). DYT it worth making so?

maxi297 reviewed (commented):
My first thoughts reading the PR description are that I'm not sure why we are focusing on those streams. They seem pretty fine if we check the source side of things:

![Screenshot 2024-05-10 at 10 09 28 AM](https://github.com/airbytehq/airbyte/assets/3360483/e449fda0-f4e7-4c91-b84f-baa00f4996e7)
[source](https://airbyte.metabaseapp.com/dashboard/3043-source-connector-stream-reliability?tab=283-connector-overview&connector_name=HubSpot&date_range=past30days&stream_name=contacts_form_submissions&connector_version=&failure_origin=&failure_type=&internal_message=&external_message=)

Based on the stream failure tab in the same dashboard, it seems like it's mostly one workspace having issues. 
![image](https://github.com/airbytehq/airbyte/assets/3360483/3d3dea34-4fbe-4cac-936f-d877074996b2)

Shouldn't we focus on destination not breaking instead? Or do we think they break because of source related issues? 

Also note that for the small percentage of source failures we see, those were mostly not read issues (see [this analysis](https://docs.google.com/document/d/1ooEl9ru7vCdW_Tu6XddNhl7fCcIZPvhWEdR844qNc14/edit#heading=h.5d9bye704goc) and [this comment](https://airbytehq-team.slack.com/archives/C03AS1GAQV6/p1713982415046919?thread_ts=1713980350.186059&cid=C03AS1GAQV6)).

I'll still review the PR but business-wise, I'm not sure this is the best this we can work on and (without having read the code so this might be misplaced fear) I fear for the complexity/risk of data loss this might induce

maxi297 commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -1374,6 +1377,62 @@ class ContactsAllBase(Stream):
+
+        next_page_token = stream_slice
+        try:
+            properties = self._property_wrapper
```
I fear this part as it sounds like it is the same logic as `Stream.read_records` and if one change, we now need to know that we need to change the other. Could we have the logic still centralize in `Stream`, have the `Stream._state` and only implement the `state` getter and setter in `ContactsAllBase`? Even then, I think we will need another class as [this](https://github.com/airbytehq/airbyte-internal-issues/issues/7674) is in progress and will rely on `ContactsAllBase` but I don't think it'll be RFR compatible as-is with both client side filtering incremental and RFR incremental...

maxi297 reviewed (commented):
In the end, I have two concerns:
* Duplication of the logic: do you think we can do something about this? 
* What happens with `contacts_list_memberships` that will be semi-incremental?

I also want to give a big thumbs up for the mock server tests which allowed me NOT to do manual testing (the only manual testing I did was to test if a vid was deleted and the API supports it very well). This is really appreciated!

maxi297 reviewed (commented):
Oh! Maybe another concern: 
* can we move most of the logic to the CDK? In theory, as long as `next_page_token` is exposed, we should be fine. I think the problem in the HubSpot case is that it re-implements `read_records` but as long as it calls for `_read_page`, we should be able to just have the change in the CDK. The reasoning is that it feels like it would build more leverage but maybe this is just an impression that needs to be confirmed

maxi297 commented on airbyte-integrations/connectors/source-hubspot/unit_tests/integrations/__init__.py:
```diff
@@ -131,3 +132,28 @@ def read_from_stream(
         return read(SourceHubspot(), cfg, cls.catalog(stream, sync_mode), state, expecting_exception)
+
+
+@freezegun.freeze_time("2024-05-05T00:00:00Z")
```
Should this only be on the actual test implementation?

maxi297 commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -1374,6 +1377,62 @@ class ContactsAllBase(Stream):
     filter_field = None
     filter_value = None
+    _state = {}
+    limit_field = "count"
```
Nice catch. It was using `limit` before. The default value on the API side is 100 so this shouldn't change the behavior

maxi297 commented on airbyte-integrations/connectors/source-hubspot/poetry.lock:
```diff
@@ -2,28 +2,31 @@
 [[package]]
 name = "airbyte-cdk"
-version = "0.78.6"
+version = "0.87.0"
```
Actually, testing manually makes me believe that upping the CDK will cause issues with `ClientSideIncrementalStream` as the state will be set and [this](https://github.com/airbytehq/airbyte/blob/1188221473e2156a591b12c655e6b853e2df65d7/airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py#L862-L864) will fail with an index error. I think it comes from a change [here](https://github.com/airbytehq/airbyte/blob/1188221473e2156a591b12c655e6b853e2df65d7/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L215) where the condition before was `if stream_state and "state" in dir(stream_instance) and <...>` but we've removed the first part

maxi297 commented on airbyte-integrations/connectors/source-hubspot/unit_tests/integrations/test_contacts_list_memberships.py:
```diff
@@ -0,0 +1,60 @@
+
+        output = self.read_from_stream(self.oauth_config(), self.STREAM_NAME, SyncMode.full_refresh)
+
+        assert len(output.records) == 12
```
Why do we expect 12 records here?

aldogonzalez8 commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -861,7 +873,10 @@ def state(self) -> Mapping[str, Any]:
     @state.setter
     def state(self, value: Mapping[str, Any]):
-        self._cursor_value = value[self.cursor_field]
+        if value:
+            self._cursor_value = value[self.cursor_field]
+        else:
+            self._cursor_value = ""
```
You can also:
```suggestion
       self._cursor_value = value[self.cursor_field] if value else ""
```

brianjlai replied:
fixed!  ty

aldogonzalez8 replied:
Awesome

brianjlai commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -861,7 +873,10 @@ def state(self) -> Mapping[str, Any]:
     @state.setter
     def state(self, value: Mapping[str, Any]):
-        self._cursor_value = value[self.cursor_field]
+        if value:
```
> Actually, testing manually makes me believe that upping the CDK will cause issues with ClientSideIncrementalStream as the state will be set and this will fail with an index error. I think it comes from a change here where the condition before was if stream_state and "state" in dir(stream_instance) and <...> but we've removed the first part

@maxi297 good comment about this in the `poetry.lock` file about CDK compatibility! So I had adjusted this condition because there's actually kind of a nasty bug in the CDK that could cause data loss for the first slice because the availability strategy can accidentally update state. See this comment in a previous [PR](https://github.com/airbytehq/airbyte/pull/37429/files#r1578493847).

And because of that change, I adjusted this condition. I think Hubspot configures setters a bit uncharacteristic from others and with RFR, we can't assume state setter/getters == incremental and will always have `cursor_field`. Basically to account for the case you mentioned, I updated our state setter to be more resilient to empty state which is what we should probably be doing regardless.

brianjlai commented:
@maxi297 sorry for the delay, switched back to low-code RFR, but back now. to answer some of your questions

> Duplication of the logic: do you think we can do something about this?
I moved the common logic into one place that both page and pages uses

> What happens with contacts_list_memberships that will be semi-incremental?
As per our discussion on Tuesday, semi-incremental and RFR won't work with each other for the moment and we don't currently have a strong need for them to right now. We can always revisit this assumption, but it would be a CDK change

> can we move most of the logic to the CDK? In theory, as long as next_page_token is exposed, we should be fine. I think the problem in the HubSpot case is that it re-implements read_records but as long as it calls for _read_page, we should be able to just have the change in the CDK. The reasoning is that it feels like it would build more leverage but maybe this is just an impression that needs to be confirmed

I am currently investigating this as an option and seeing how we can get this "for free" for more connectors. My initial thought is that we can't for Hubspot because while it does implement `HttpStream.next_page_token`, it doesn't actually use the default `HttpStream.read_records()` which is where in the CDK I would try to add the automatic RFR logic. I will investigate this more as I try to vet the hypothesis. If we can get for free I can always clean out Hubspot, but I'd like to try to get RFR out for Hubspot. The RFR for free implementation if feasible would still be a few weeks out

Otherwise I made most of the other PR suggestions and incorporated your latest set of PRs which adjusted my mock server test implementations to be more in line w/ the review comments

maxi297 reviewed (approved):
Just nits. LGTM! Thanks Brian

aldogonzalez8 reviewed (approved):
APPROVED

brianjlai commented on airbyte-integrations/connectors/source-hubspot/unit_tests/integrations/test_contacts_merged_audit.py:
```diff
@@ -0,0 +1,128 @@
+_VID_OFFSET = 5331889818
+
+
+class TestContactsFormSubmissionsStream(TestCase, HubspotTestCase):
```
oops yeah copy paste forgot, thanks!

brianjlai commented on airbyte-integrations/connectors/source-hubspot/unit_tests/integrations/test_contacts_merged_audit.py:
```diff
@@ -0,0 +1,128 @@
+            )
+        ]
+
+        # Even though we only care about the request with a vidOffset parameter, we mock this in order to pass the availability check
```
added that to both my tests. thanks!

brianjlai commented:
<img width="1656" alt="hubspot_live_test" src="https://github.com/airbytehq/airbyte/assets/51336873/f0545fce-f133-42e5-813e-4f889a345ef4">

passing live test run
